text/0000-try-expr.md

```diff
@@ -218,18 +233,22 @@ This is our choice of keyword, because it:
 
 1. has a massive dominance in both popular and less known languages and is
    sufficiently semantically faithful to what `try` means in those languages.
+   Thus, we can leverage people's intuitions and not spend too much of our
+   complexity budget.
 2. is consistent with the standard library wrt. `Try` and `try_` prefixed methods.
-3. it is brief.
-4. it has high fidelity wrt. the concepts it attempts to communicate (exception boundary for `?`).
+3. it is brief. 
+4. it has high fidelity wrt. the concepts it attempts to communicate
+   (exception boundary for `?`). This high fidelity is from the perspective of
+   a programmers intent, i.e: "I want to try a bunch of stuff in this block".
 5. it can be further extended with `catch { .. }` handlers if we wish.
 
 ## Alternative: reserving `catch`
 
-1. **Fidelity to the construct's actual behavior:** Low
-2. **Precedent from existing languages:** None
+1. **Fidelity to the construct's actual behavior:** High
+2. **Precedent from existing languages:** Erlang and Tcl, see [prior-art]
 3. **Brevity / Length:** 6
-4. **Consistency with the naming of the trait used for `?`:** Inconsistent
-5. **Consistency with related libstd fn conventions:** Somewhat consistent
+4. **Consistency with related libstd fn conventions:** Somewhat consistent
```

Comment:
The [only use of `catch` in std](https://doc.rust-lang.org/std/?search=catch) is for `catch_unwind` which is technically unrelated and may even be misleading into thinking that `catch {}` catches panics.

Reply:
Mmh, yeah; so the `catch_unwind` is why I wrote "somewhat consistent"; but your reasoning is compelling. I should probably change this to "Misleading", right?

Reply:
idk, it's not really that misleading -- it says that it's catching an unwind. It's still very consistent in that it talks about the operation "catch" in the same way.

Reply:
How are they unrelated? Aren't they both catching things?

Reply:
Just as both `Result<T>` and `Wrapping<T>` are wrapping things :)

`catch {}` is not catching panics, even though "catching" and "panics" go together in stdlib. And `catch_unwind` doesn't catch "unwinding" via `?`. So given two kinds of errors and separate ways of propagating them in Rust, there's risk of confusing which is which if two technically very different and incompatible constructs share a name.

Reply:
Panicking is hardly related to short-circuited `try` blocks. They are entirely different things.

Reply:
To be honest, this is a bit frustrating. `catch` was also proposed as a future extension to the proposed `try`. People also keep referring to `?` as throwing. There's also the argument that `try`/`catch` are good terminology for Rust because people know it and it's close enough. But `catch` can't be used for catching stuff "thrown" by `?` because another thing exists that stops panics from propagating and that isn't catching even though that is closer to exceptions?

This back and forth over semantics of words and how things fit and don't fit and fit enough or not enough makes me again think that the best course of action would be to stay clear of `try`/`catch` completely.

Reply:
I know, it's very unfortunate that stdlib uses `try_*` for `Result` and `catch` for unwinding. However, within Rust's terminology it's still possible to keep them consistent. Or pretend `catch_panic` doesn't exist and try not to talk about "catching" panics.

Maybe `wrap {}` vs `.unwrap()` is not that bad as an alternative?

Reply:
Some 🚲🏠 is ongoing on what to call the `?` operator, see: https://internals.rust-lang.org/t/bikeshed-a-consise-verb-for-the-operator/7289/19

Reply:
We already have lots of `try_xxx()` functions in std and across crates.io, most of which evaluate to `Result<T, E>` (just like the proposed *block* construct would). Why not accept this correspondence?

I don't see how `try` is problematic. Try is the perfect verb for the undertaking of actions that are fallible, which is why it is used as prefix for these fn names, *and* the proposed fallible block construct.

Reply:
I have been told those functions aren't actually trying...

For me it is problematic because of it's connection with exception handling in other languages (I understand this is the appeal for some).

Control flow and error handling are simply two of the points where I would appreciate a solution that isn't ambiguous, that doesn't attempt to look like something else with some differences, that doesn't invoke associations with foreign concepts. Not just for this, or the upcoming edition changes, but also for code that is 10 years out as well.

I've made (I feel like) many comments above about my issues with it. If I was unable to communicate *any* issues about `try`, I must've done a bad job.

Reply:
No one is disputing that there are drawbacks with `try`, and indeed the RFC has some of those listed; I believe we differ on how we value the trade-offs with respect to benefits and drawbacks.

Reply:
"Connection with exception handling in other languages" is a valid point, and I think everyone agrees that Rust's handling is technically different from "traditional" try/catch.

However, just that argument isn't either for or against renaming `catch` to `try`, because *both* are connected to the same terminology.

If we agree that avoiding other languages terms is the # 1 priority (over familiarity, backwards compatibility with Rust's own existing terminology), then we need to have a strong contender for the alternative name for the block, and for `try_` functions. I think if we rename one, but not the other, we'll have worst of both worlds — terminology inconsistent with other languages, and terminology inconsistent within Rust.

Reply:
I don't really think that's true. `try_foo` functions are describing what they do; it's not in an attempt to draw a connection with exception based languages.

In fact, as mentioned in this RFC thread often, `try_foo` uses `try` in the exact opposite sense as the proposed `try {}`. IMO it would be totally consistent to use a new keyword for `try {}`/`catch {}` but continue to use `try_*`. `try_*` isn't syntax, it's a naming convention for functions, and as such doesn't invoke parallels as strongly as a block named `try {}`.

Reply:
@Centril I'm specifically responding to repax's "I don't see how try is problematic."

@kornelski I'm not too bothered by the method names, since they are used for a wide variety of things (optionals, failables, non-blocking variants, non-panicky versions, and so on), which is another reason I find the term too overloaded for a syntax construct. But I certainly wouldn't be opposed to better terminology for the individual cases.

And to reiterate my position: I feel `catch` is closer than `try` because, semantically, I feel the construct is closer to the interception than the actual attempt. A function returning a `Result<..., ...>` can still `?` things and thus try them, but you'd use this block to locally intercept and handle things.

Reply:
I feel like despite the technical differences between `try {}` and `try_*` functions, the semantics of both are "You're going to get a `Result` from this thing." To me, the semantics are consistent in that way.

Reply:
Instead of `try` or `catch` we could find another name for the block construct.

Consider this: a `fn` and a `closure` are both perfectly capable of capturing the errors from `?` without requiring any special keyword. The block construct could have a name that make you think more of a code routine with a (return) value. Perhaps `proc`? It's already a reserved keyword.

Reply:
@ErichDonGubler I agree with this sentiment. As I noted above some time ago:

> In a sense yes; Remembering that `try { .. }` is an expression that is usually assigned to a binding: `let foo = try { .. }`, then you are trying some stuff and saving the result in `foo`. Equivalently, function application with `let foo = fallible();` (assuming fallible : `fn() -> Result<T, E>`) is trying and saving to `foo`. Therefore function application for `-> Try<Ok = T, Err = E>` is the `try { .. }` of sorts.

TL;DR: Function application of a `try_` function acts like `try { .. }` does semantically but on a function level instead of scope level.

Reply:
@ErichDonGubler:

> I feel like despite the technical differences between try {} and try_* functions, the semantics of both are "You're going to get a Result from this thing." To me, the semantics are consistent in that way.

You may not always going to get a `Result`. If the `TryContinue`, `Try`, `TryBreak` design is accepted, then breaking bubbles the type implementing `Try`. That is *typically* a result, but the semantics are closer to @phaylon's "just another value".

That said, those methods do follow the same semantics as `try { .. }`, as noted by @Centril, as the return value of a `try_` function typically implements `Try`. The only exception I found in `std` is `std::sync::mpsc::Receiver::try_iter`, which produces an iterator that "tries" when it advances ([TryIter](https://doc.rust-lang.org/std/sync/mpsc/struct.TryIter.html).